src/transforms/geoip.rs

```diff
@@ -0,0 +1,256 @@
+                if let Some(continent_code) = continent_code {
+                    lookup_results.insert(Atom::from("continent_code"), continent_code.into());
+                }
+
+                let iso_code = data.country.and_then(|cy| cy.iso_code);
+                if let Some(iso_code) = iso_code {
+                    lookup_results.insert(Atom::from("country_code"), iso_code.into());
+                }
+
+                let time_zone = data.location.clone().and_then(|loc| loc.time_zone);
+                if let Some(time_zone) = time_zone {
+                    lookup_results.insert(Atom::from("time_zone"), time_zone.to_string());
+                }
+
+                let latitude = data.location.clone().and_then(|loc| loc.latitude);
+                if let Some(latitude) = latitude {
+                    lookup_results.insert(Atom::from("latitude"), latitude.to_string());
+                }
+
+                let longitude = data.location.clone().and_then(|loc| loc.longitude);
+                if let Some(longitude) = longitude {
+                    lookup_results.insert(Atom::from("longitude"), longitude.to_string());
+                }
+
+                let postal_code = data.postal.clone().and_then(|p| p.code);
+                if let Some(postal_code) = postal_code {
+                    lookup_results.insert(Atom::from("postal_code"), postal_code.into());
+                }
+            }
+
+            let geoipdata = GeoipDecodedData {
+                city_name: lookup_results
+                    .get(&Atom::from("city_name"))
+                    .unwrap_or(&String::from(""))
+                    .to_string(),
+                continent_code: lookup_results
+                    .get(&Atom::from("continent_code"))
+                    .unwrap_or(&String::from(""))
+                    .to_string(),
+                country_code: lookup_results
+                    .get(&Atom::from("country_code"))
+                    .unwrap_or(&String::from(""))
+                    .to_string(),
+                time_zone: lookup_results
+                    .get(&Atom::from("time_zone"))
+                    .unwrap_or(&String::from(""))
+                    .to_string(),
+                latitude: lookup_results
+                    .get(&Atom::from("latitude"))
+                    .unwrap_or(&String::from(""))
+                    .to_string(),
+                longitude: lookup_results
+                    .get(&Atom::from("longitude"))
+                    .unwrap_or(&String::from(""))
+                    .to_string(),
+                postal_code: lookup_results
+                    .get(&Atom::from("postal_code"))
+                    .unwrap_or(&String::from(""))
+                    .to_string(),
+            };
+            event.as_mut_log().insert_explicit(
+                Atom::from(self.target.clone()),
+                serde_json::to_string(&geoipdata).unwrap().into(), //FIXME: handle pnic heere
```

Comment:
Instead of storing these in an intermediate map and serializing it to a json string, how about we insert them into the event as we go? We could namespace the keys under something configurable like `geoip.` to avoid conflicts.

Reply:
Hi @lukesteensen i was under the impression that we wanted to have the geoip decoded results nested inside a user specified field? That is, something like this:

```
{ 
"key":" value",
"geoip": {
    "city": "..",
    "countrycode": ".. 
 }
}
```

Are you suggesting we instead have `geoip.city` as top level fields?

Reply:
Yes. Currently, vector doesn't really support nested fields. We're looking to change this soon, but right now this PR adds one top-level field that holds a json string, not nested fields.

Until we support real nested fields, we've been "flattening" keys. So the closest to your example json would be a `geoip.city` field, `geoip.countrycody` field, etc.